### ISSUE TITLE
The cursor isn't hidden right away

### ISSUE THREAD

**Issue by arcanis** [mintty/mintty#612]
I'm printing a string similar to the following:

```
"\u001b[?25l\u001b[68;1H <about a thousand additional characters> \u001b[42;1H\u001b[34h\u001b[?25h"
```

Which should be interpreted as "hide the cursor, move the cursor to 68;1, print the additional characters, move back the cursor to 42;1, then show the cursor". My goal is to print anywhere on the screen without the cursor to blink (or very little).

Unfortunately, it seems that mintty ignore for a short time the request to hide the cursor since, when I test this, I actually see the cursor moving. It's very fast, so most of the time it just appear on another line/column for only a few milli-micro-seconds, but it's still noticeable that the cursor actually moved.

Is there anything I can do to prevent this from happening? I'm not aware of any terminal sequence that would allow the terminal to "buffer" my commands and execute them all at once, but I haven't much experience with terminal sequences yet.

**Comment by mintty** [comment]
Is the effect visible if you disable cursor blinking?

**Comment by arcanis** [comment]
Yes, it still appears

**Comment by mintty** [comment]
I do not reproduce any disturbing visual artefact. Please describe a more precise test case (exact test output, screen size, font and font size, maybe full config file (reduced if possible)).

**Comment by arcanis** [comment]
Here is a video of the effect and a ttyrecord (use `ttyplay <path/to/ttyrecord>` to run it): [tty.zip](https://github.com/mintty/mintty/files/637831/tty.zip)

My screen size for this test is 80x13, and my settings are the following:
![image](https://cloud.githubusercontent.com/assets/1037931/20988101/883a128e-bccf-11e6-9b10-da87b006b79b.png)

**Comment by mintty** [comment]
Please provide an output log (`mintty -l`) or where would I get that ttyplay tool?

**Comment by arcanis** [comment]
It's available as the `ttyrec` package on standard ubuntu - I'm not sure if it is available on cygwin :/

**Comment by dschnee** [comment]
I've also seen this. For the above ttyrecord it is noticeable if the terminal window is small or if running in tmux.

**Comment by mintty** [comment]
I've traced the part of code that's supposed to draw the cursor, wintext.c:win_text: if (has_cursor) ...,
and it's never invoked while the cursor is hidden.
However, if I disable that code completely, there's still a visible cursor... Puzzled.

**Comment by mintty** [comment]
What's actually the idea of ESC[34h? There is no such sequence.

**Comment by arcanis** [comment]
I failed my initial copy-paste. `ESC[34h` is actually preceding `ESC[?25h` (`ESC[34hESC[?25h` is the `cnorm` instruct than re-display the cursor).

**Comment by arcanis** [comment]
Couldn't there be a race condition were you would queue multiple display, but instantly evaluate the cursor display flag? So for example you would hide the cursor, queue a display, show the cursor, and then the display would be printed, with its cursor still on?

**Comment by mintty** [comment]
`ESC[34h` is not defined in mintty or xterm. Where do you get that from?
Yes, I've been looking for a race condition, especially as mintty display refresh is timer-controlled (`do_update`). But I don't see any hint to such an inconsistency (yet).

**Comment by arcanis** [comment]
Apparently it's from the `screen` terminfo:

```
$> TERM=screen tput cnorm | cat -e
^[[34h^[[?25h%
```

**Comment by mintty** [comment]
I cannot reproduce the issue although I see the artefact in the video.
Please try the current repository version and enable `debug_cursor` in wintext.c; for me, it nether shows that the cursor would be displayed in an inappropriate line while it's hidden.
If we get a reproducible test environment, maybe I can do something, whether race condition or not...

**Comment by mintty** [comment]
Which version of mintty do you run, actually?
I've retested on a slow machine and can still not reproduce it.

**Comment by arcanis** [comment]
I'm currently using wsltty 2.7.1

**Comment by arcanis** [comment]
@mintty I've gotten around compiling mintty, can I use my custom build with wsltty? I've tried replacing the `mintty.exe` file, but unfortunately it crashed (0xc000007c). Should I copy a DLL somewhere?

**Comment by arcanis** [comment]
I haven't been able to reproduce this glitch in the last release, *except when running inside tmux*. Maybe tmux does some kind of buffering and send those commands in multiple parts?

**Comment by arcanis** [comment]
Here are my logs:


```
do_update cursor_on 1 @22,4
do_update cursor_on 1 @22,4
do_update cursor_on 1 @22,4
do_update cursor_on 1 @1,23
set cursor (colour BFBFBF) @(row 1 col 23) cursor_on 1
painting cursor_type 'b' cursor_on 1
do_update cursor_on 1 @22,4
set cursor (colour BFBFBF) @(row 22 col 4) cursor_on 1
painting cursor_type 'b' cursor_on 1
do_update cursor_on 1 @22,4
```

As you can see, the cursor briefly jumped to `1,23`.

**Comment by mintty** [comment]
The mintty build for wsltty uses cygwin 32-bit, that may be reason for your initial problems.
Were you using tmux originally? Maybe it doesn't even support the hide cursor sequence?

**Comment by arcanis** [comment]
Hm, tmux does support the hide cursor sequence (it correctly hides the cursor when running `tput civis`). That's very weird.

**Comment by mintty** [comment]
Is this still an issue?

**Comment by mintty** [comment]
I've uploaded a tentative fix. Please check.

**Comment by mintty** [comment]
Released 2.8.5, with the patch. Please check the new version for the issue.

**Comment by mintty** [comment]
Closing the issue on the assumption that the patch may have resolved the issue, and I could never reproduce it myself.
